shark_turbine/kernel/wave/decompose_reduce_ops.py

```diff
@@ -0,0 +1,112 @@
+from ..wave.constraints import (
+    Constraint,
+    HardwareConstraint,
+    WorkgroupConstraint,
+    TilingConstraint,
+)
+from .._support.tracing import CapturedTrace
+from .._support.indexing import IndexingContext, IndexSequence, IndexSymbol, IndexExpr
+from ..ops.wave_ops import get_custom, Add, ReduceOp, ShuffleOp, CustomOp, ExtractSlice
+
+from .utils import DCE
+import torch.fx as fx
+import math
+from typing import Callable
+
+TKW_COMBINER = {"sum": Add}
+
+
+def get_graph_node(custom: CustomOp, graph: fx.Graph):
+    custom.add_to_graph(graph)
+    custom = custom.fx_node
+    return custom
+
+
+def emit_local_reduction(
+    binary_fn: Callable, src: fx.Node, graph: fx.Graph, local_reduction_size: int
+) -> fx.Node:
+    init = get_graph_node(ExtractSlice(src, [0], [1], [1]), graph)
+    for i in range(1, local_reduction_size):
+        cur_slice = get_graph_node(ExtractSlice(src, [i], [1], [1]), graph)
+        init = get_graph_node(binary_fn(init, cur_slice), graph)
+    return init
+
+
+def emit_global_reduction(
+    binary_fn: Callable, src: fx.Node, graph: fx.Graph, subgroup_size: int
+) -> fx.Node:
+    init = src
+    num_steps = int(math.log2(float(subgroup_size)))
+    for i in range(num_steps):
+        shuffle_offset = 2**i
+        shuffle_val = ShuffleOp(init, shuffle_offset, subgroup_size)
+        shuffle_node = get_graph_node(shuffle_val, graph)
+        init = get_graph_node(binary_fn(init, shuffle_node), graph)
+    return init
+
+
+def decompose_reduce_ops(
+    trace: CapturedTrace,
+    constraints: list[Constraint],
+    index_map: dict[IndexSymbol, int],
+):
+    """
+    The lowering for multi_reduction is done in two steps:
+      1. Local Reduce: Each thread reduces all elements carried by it along
+         the reduction dimensions.
+      2. Thread Reduce: Each thread reduces result of step 1 across threads
+         by doing a butterfly shuffle.
+      3. Accumulator Reduce: Each thread reduces it's intermediate reduced
+         results with the accumulator it holds.
+    """
+    # Get reducte nodes.
+    reduce_nodes = trace.walk(lambda node: isinstance(get_custom(node), ReduceOp))
+    if not reduce_nodes:
+        return
+
+    # Setup constraints
+    hardware_constraint = next(
+        c for c in constraints if isinstance(c, HardwareConstraint)
+    )
+    constraint_tile_size = {
+        c.dim: c.tile_size
+        for c in constraints
+        if isinstance(c, TilingConstraint) or isinstance(c, WorkgroupConstraint)
+    }
+    subgroup_size = hardware_constraint.threads_per_wave
+    for node in reduce_nodes:
+        custom = get_custom(node)
+        with custom.graph.inserting_before(custom.fx_node):
+            reduction_src, reduction_acc, reduction_dim = node.args
+            binary_fn = TKW_COMBINER[custom.tkw_op_name]
+            if reduction_dim is None:
+                reduction_dim = custom.type.symbolic_shape[-1]
```

Comment:
As I mentioned in other comment, when dimension is not specified, user would expect reduction over all dimensions instead of fastest (following numpy rules). For now, let's just bail out and ask user to explicitly specify dimension.

Reply:
Makes sense, done, thanks! :)